nativelink-store/src/store_manager.rs

```diff
@@ -22,18 +24,56 @@ use parking_lot::RwLock;
 pub struct StoreManager {
     #[metric]
     stores: RwLock<HashMap<String, Store>>,
+    store_config_anti_collision_digests: RwLock<Vec<String>>,
 }
 
 impl StoreManager {
     pub fn new() -> StoreManager {
         StoreManager {
             stores: RwLock::new(HashMap::new()),
+            store_config_anti_collision_digests: RwLock::new(vec![]),
         }
     }
 
-    pub fn add_store(&self, name: &str, store: Store) {
+    pub fn add_store(&self, name: &str, store: Store) -> Result<(), Error> {
         let mut stores = self.stores.write();
-        stores.insert(name.to_string(), store);
+
+        if stores.contains_key(name) {
+            return Err(make_err!(
+                Code::AlreadyExists,
+                "a store with the name '{}' already exists",
+                name
+            ));
+        }
+
+        for existing_store in stores.values().into_iter() {
+            if ptr::eq(&store, existing_store) {
```

Comment:
Will remove this check. Thanks to rust's memory model and the fact that in this design, ownership of the struct instance is passed to the StoreManager on line 58. We don't have Store instance pointer being passed around with a risk of passing the same pointer N times. We have a single instance of which ownership is transferred exactly once in guaranteed fashion (thanks Rust).

So instance protection is no longer a goal or needed. But I do think the config overlap detection is valuable.

Reply:
Thanks Rust is an underrated comment.